src/pipecat/services/aws/tts.py

```diff
@@ -247,4 +249,28 @@ def read_audio_data(**args):
             yield ErrorFrame(error=error_message)
 
         finally:
+            if is_tracing_available():
+                from opentelemetry import trace
+
+                from pipecat.utils.tracing.helpers import add_tts_span_attributes
+
+                current_span = trace.get_current_span()
+                service_name = self.__class__.__name__.replace("TTSService", "").lower()
+
+                ttfb_ms = None
+                if hasattr(self._metrics, "ttfb_ms") and self._metrics.ttfb_ms is not None:
+                    ttfb_ms = self._metrics.ttfb_ms
+
+                add_tts_span_attributes(
+                    span=current_span,
+                    service_name=service_name,
+                    model=self._settings.get("engine", "standard"),
+                    voice_id=self._voice_id,
+                    text=text,
+                    settings=self._settings,
+                    character_count=len(text),
+                    operation_name="tts",
+                    ttfb_ms=ttfb_ms,
+                )
```

Comment:
I'm wondering if we could create annotations for all this so we don't have to write these things every time.

```
    @tracing(name="polly_tts")
    async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
```
@tracing would call:

```
    @traced(attachment_strategy=AttachmentStrategy.CHILD, name="polly_tts")
```

One concern I have is that this is OpenTelemetry specific. I think we should make it so you can replace the service you want to use for telemetry in an easy way.

Reply:
As far as I could understand, OpenTelemetry is an open-source framework whose goal is to unify tracing, metrics, and logs into a single standard.

And it looks like most vendors support the OpenTelemetry Protocol: https://opentelemetry.io/ecosystem/vendors/

And even in cases where that’s not possible, we should be able to write a custom exporter to create anything custom and send it to any destination.

So, I don't think it would be bad to rely on OpenTelemetry for that. But maybe I am still missing something.

Is there any particular reason why are you concerned about being OpenTelemetry specific for structuring and collecting our traces ?

Reply:
We discussed this with @markbackman . I thought it was cleaner because you can then have more control in your `@tracing` decorator which is what Mark ended up implementing. For example, you can have turn-based spans, etc. Also, as I will comment in the other thread, it's possible that we might to do something different than OpenTelemetry, I'll add more details there.